flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -89,17 +96,15 @@
 	/** The number of available buffers that have not been announced to the producer yet. */
 	private final AtomicInteger unannouncedCredit = new AtomicInteger(0);
 
-	/**
-	 * The latest already triggered checkpoint id which would be updated during
-	 * {@link #spillInflightBuffers(long, ChannelStateWriter)}.
-	 */
-	@GuardedBy("receivedBuffers")
-	private long lastRequestedCheckpointId = -1;
+	private final BufferManager bufferManager;
 
-	/** The current received checkpoint id from the network. */
-	private long receivedCheckpointId = -1;
+	/** Stores #overtaken buffers when a checkpoint barrier is received before task thread started checkpoint. */
+	@GuardedBy("receivedBuffers")
+	private Map<Long, Integer> numBuffersOvertaken = new HashMap<>();
 
-	private final BufferManager bufferManager;
+	/** All started checkpoints where a barrier has not been received yet. */
+	@GuardedBy("receivedBuffers")
+	private Deque<Long> pendingCheckpointBarriers = new ArrayDeque<>(2);
```

Comment:
I don't think it's worth complicating the code with support for multiple concurrent checkpoints. It's not likely to be implemented soon, if ever.

Reply:
Changed it to support only one concurrent checkpoint. I also extracted the whole logic into one helper class that can be used by both `LocalInputChannel` and `RemoteInputChannel` (with syncrhonization).